homeassistant/components/lock/xiaomi_aqara.py

```diff
@@ -0,0 +1,91 @@
+"""
+Support for Xiaomi Aqara Lock.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/lock.xiaomi_aqara/
+"""
+import asyncio
+import logging
+from homeassistant.components.xiaomi_aqara import (PY_XIAOMI_GATEWAY,
+                                                   XiaomiDevice)
+from homeassistant.components.lock import LockDevice
+from homeassistant.const import (STATE_LOCKED, STATE_UNLOCKED)
+from homeassistant.helpers.event import async_call_later
+
+_LOGGER = logging.getLogger(__name__)
+
+FINGER_KEY = 'fing_verified'
+PASSWORD_KEY = 'psw_verified'
+CARD_KEY = 'card_verified'
+VERIFIED_WRONG_KEY = 'verified_wrong'
+
+ATTR_VERIFIED_WRONG_TIMES = 'verified_wrong_times'
+
+UNLOCK_MAINTAIN_TIME_S = 5
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Perform the setup for Xiaomi devices."""
+    devices = []
+
+    for gateway in hass.data[PY_XIAOMI_GATEWAY].gateways.values():
+        for device in gateway.devices['lock']:
+            model = device['model']
+            if model == 'lock.aq1':
+                devices.append(XiaomiAqaraLock(device, 'Lock', gateway))
+    add_devices(devices)
+
+
+class XiaomiAqaraLock(LockDevice, XiaomiDevice):
+    """Representation of a XiaomiAqaraLock."""
+
+    def __init__(self, device, name, xiaomi_hub):
+        """Initialize the XiaomiAqaraLock."""
+        self._changed_by = 0
+        self._verified_wrong_times = 0
+
+        super().__init__(device, name, xiaomi_hub)
+
+    @property
+    def is_locked(self) -> bool:
+        """Return true if lock is locked."""
+        if self._state is not None:
+            return self._state == STATE_LOCKED
+
+    @property
+    def changed_by(self) -> int:
+        """Last change triggered by."""
+        return self._changed_by
+
+    @property
+    def device_state_attributes(self) -> dict:
+        """Return the state attributes."""
+        attributes = {
+            ATTR_VERIFIED_WRONG_TIMES: self._verified_wrong_times,
+        }
+        return attributes
+
+    @asyncio.coroutine
```

Comment:
Could you care about this one, too?

Reply:
I'm not sure for this comment.
This function will be invoked in async_call_later, I need this function called after UNLOCK_MAINTAIN_TIME seconds. but callback function can not do IO operation, I think delay should not called in callback too.

Reply:
`async_call_later` is a callback.

Reply:
I try to change this by using async_run_job, I tested this can work:
```
@@ -67,8 +68,10 @@ class XiaomiAqaraLock(LockDevice, XiaomiDevice):
         return attributes
 
     @asyncio.coroutine
-    def clear_unlock_state(self, _):
+    def clear_unlock_state(self):
         """Clear unlock state automatically."""
+        yield from asyncio.sleep(UNLOCK_MAINTAIN_TIME)
+
         self._state = STATE_LOCKED
         self.async_schedule_update_ha_state()
 
@@ -85,8 +88,7 @@ class XiaomiAqaraLock(LockDevice, XiaomiDevice):
                 self._changed_by = int(value)
                 self._verified_wrong_times = 0
                 self._state = STATE_UNLOCKED
-                async_call_later(self.hass, UNLOCK_MAINTAIN_TIME,
-                                 self.clear_unlock_state)
+                self.hass.async_run_job(self.clear_unlock_state)
                 return True
 
         return False

```

But I failed to using *@callback*
I have tried this 
```
@@ -66,9 +67,12 @@ class XiaomiAqaraLock(LockDevice, XiaomiDevice):
         }
         return attributes
 
+    @callback
     @asyncio.coroutine
-    def clear_unlock_state(self, _):
+    def clear_unlock_state(self):
         """Clear unlock state automatically."""
+        yield from asyncio.sleep(UNLOCK_MAINTAIN_TIME)
+
         self._state = STATE_LOCKED
         self.async_schedule_update_ha_state()
 
@@ -85,8 +89,7 @@ class XiaomiAqaraLock(LockDevice, XiaomiDevice):
                 self._changed_by = int(value)
                 self._verified_wrong_times = 0
                 self._state = STATE_UNLOCKED
-                async_call_later(self.hass, UNLOCK_MAINTAIN_TIME,
-                                 self.clear_unlock_state)
+                self.hass.async_run_job(self.clear_unlock_state)
                 return True
 
```
But ```clear_unlock_state(self)``` will not invoked.

I'm sorry that how to use callback decorator, can you point me out how to do this?

Reply:
Hi @MartinHjelmare  @syssi 
May you help me to improve this code?
I have checked the code in homeassistant, which used async_call_later and callback, but do not find a similar usage in my situation.

Reply:
Done.
Thank you for your advice